cupy/cuda/memory.pyx

```diff
@@ -419,44 +419,49 @@ class PooledMemory(Memory):
         self.size = chunk.size
         self.pool = pool
 
-    def __del__(self):
-        if self.ptr != 0:
-            self.free()
-
     def free(self):
         """Frees the memory buffer and returns it to the memory pool.
 
         This function actually does not free the buffer. It just returns the
         buffer to the memory pool for reuse.
 
         """
+        cdef Py_ssize_t ptr
+        ptr = self.ptr
+        if ptr == 0:
+            return
+        self.ptr = 0
         pool = self.pool()
```

Comment:
I am still curious for a situation that `self.pool` (weakref) can be None or not.
I mean that codes like below would be required.

```python
if self.pool is None:
    return
pool = self.pool()
if pool is None:
    return
```

Reply:
I I have never encountered a case where the self variable automatically becomes `None`.
Do I need to fix this code?

Reply:
I am not sure, but sounds okay without fixing.